### DISCUSSION TITLE
New discussions, new maintainer

### DISCUSSION THREAD

**Discussion by emcniece** [cncjs/cncjs#749] [Announcements]
## 👋 Hey, new maintainer onboarding.

Hi everyone. @cheton and @MitchBradley have warmly accepted my offer to help out with maintainership of the CNCjs project. It'll take me a little while to get up to speed, but to start I'm focusing on triaging the issue backlog and consolidating the CI pipeline so that we can start moving tickets through the queue. You might see your issue get labelled or assigned to a Project, that's likely me getting organized!

One of the major pain points I've heard about so far is the variance in deploy platforms and how tricky they can be. I'll keep my eyes open for these tickets in the queue. I'm also interested in optimization, and while that will come later I've started to peek into the Webpack bundle and the Docker image.

Here's a treemap of the production Webpack bundle (~15Mb):

<img width="1439" alt="Screen Shot 2022-03-25 at 9 17 32 PM" src="https://user-images.githubusercontent.com/1182187/160224394-7a409f2f-b55e-4484-b0f9-81609cbdcb9e.png">

And here's an analysis of the most recently published Docker image (~3Gb uncompressed):

<img width="1440" alt="Screen Shot 2022-03-25 at 9 18 14 PM" src="https://user-images.githubusercontent.com/1182187/160224413-a83e531b-4858-4b62-a814-fd5b98d72e80.png">

I look forward to digging in to this work over the next few months. Here's a rough timeline:

- +2 weeks: organize the 199 open issues
- +1 month: publish a new release
- +2 months: migrate CI pipelines to Github Actions
- +4 months: distribution optimization

What's your favourite feature of CNCjs? What gives you the most grief? I'm all ears.

**Comment by neilferreri** [comment]
@emcniece Welcome and thank you for picking this up.

The biggest issue I know of is the random hangups, especially on larger files. There have been some issues here, but it's more common to see things posted on CNC user forums. Sometimes, the GUI will be non responsive but the job continues without issue and sometimes it all just needs to be restarted. 
Also, if you haven't already, please take a look at gSender by Sienci Labs which is based on cncjs for implementation of some new features.

**Reply by emcniece** [reply]
Thanks for the quick input! Out of curiosity, do you have any links to said CNC forums handy?

**Reply by cheton** [reply]
The Node.js server program is fine to handle 500MB or more. However, the browser itself cannot render large G-code files (said 20MB or more) because of heavy CPU & memory consumption for 3D modeling and computation.

The Snapmaker team (https://github.com/Snapmaker/Luban) may have resolved this issue using a different way, I need to check if there is an ideal solution to overcome the UI issue.

**Comment by MitchBradley** [comment]
Facebook CNCjs Users Group is the primary one.

**Comment by neilferreri** [comment]
I don't do Facebook, but the biggest hobby level groups in the US are probably Carbide 3D and Inventables. Both have user forums and both have "unofficial" Facebook groups. 

Here's a link to one of the issues. I actually saw a new one raised a day or so ago that I believe you responded to as possibly being an EMI issue. This is a different thing... definitely a cncjs issue.
[Issue 394](https://github.com/cncjs/cncjs/issues/394)

**Reply by emcniece** [reply]
The Carbide3D forums were indeed where I first heard about CNCjs!

**Comment by emcniece** [comment]
Just finished importing the open issues to a new project: https://github.com/orgs/cncjs/projects/3/views/1?layout=table

**Reply by cheton** [reply]
It looks great!

**Comment by saxa** [comment]
@emcniece welcome, and well done. I am also here ready to do more in here, I want to mainly see if I can deply this on my machines 
on a slackware os, if I get packages ready for install and other stuff like scripts I will post on the wiki. 
I do not care about forums as I preffer to have things in bug trackers sorted. 
Also I do not use Facebook, but a general forum would be nice, although I think on forums things get quickly lost :)

**Reply by emcniece** [reply]
Thank you for the offer! I've seen Mitch call this out already: documentation efforts are most appreciated. Let us know how it goes.

**Reply by MitchBradley** [reply]
Please note that just adding instructions to the wiki, without going back and seeing how it relates to other instructions that are already there and possibly fixing them, just makes the overall problem worse.  It is hard to know where a given new user will "land" in the instruction space.  The existence of a lot of different sets of instructions confuses users.
What we really need is for someone to go through all of the many conflicting sets of instructions, both on the wiki and cncjs.org, delete a bunch of them, get rid of stale information, and make a small, tidy, coherent set.

**Reply by saxa** [reply]
I do agree about the lots of different instructions on the wiki, this surely is confusing, especially for people who do not like to take a good look. My intention is to document what I done , dispose the correct instruction for the platform I use. But I can do this also on some other platform, that is not a problem. I also do think that one of the purposes of documentation is to have even old obsolete stuff there at disposal for who needs it. I do not see that as a problem until it is organized and clear.

**Comment by MitchBradley** [comment]
I understand the desire to avoid Facebook, but doing so effectively isolates one from the vast majority of users, who don't know how to use GitHub issues.  That isolation is bad from the standpoint of community support, but can be good for one's personal sanity.

**Reply by saxa** [reply]
Yeah, I personally am part of that group iirc, but never had received much notifications from there about peoples posts. Simply I see FB as a different tool, which has nothing much to do with the way which could be much helpful for this project. Also a lot of information be in different places just make it harder to find it. FB has some kind of protection from other search engines, some other sites like to ask a registration just to see the posts etc... Anyway, nothing against it, but I would prefer to see if people would be directed to some ticket here to discuss the topic there instead of FB.Again thisis my opinion and I have nothing against some of the information be on FB.

**Comment by MitchBradley** [comment]
Consider what happens when a user searches for, say, "cncjs installation".  Who knows which page they will land on?  If they are new to CNCjs they will not understand any of the context necessary to determine whether the instructions they are seeing are relevant to their situation, whether they are current or hopelessly out of date, or any number of other factors that could lead them down a rathole.  Ideally, every page with instructions would start with a link back to an overview that gives context about how to proceed in different situations, and then describes the context and limitations of the information on the current page.

Unfortunately, that level of context is rarely provided.  When someone creates a new page of instructions, their typical mindset is that they are deeply immersed in what they are doing at the moment.  What they write reflects that narrow focus, and then when a new user lands there, lacking any of that context, they are utterly confused.

Good documentation is hard.  Really hard.

**Reply by saxa** [reply]
`Consider what happens when a user searches for, say, "cncjs installation". Who knows which page they will land on? If they are new to CNCjs they will not understand any of the context necessary to determine whether the instructions they are seeing are relevant to their situation, whether they are current or hopelessly out of date, or any number of other factors that could lead them down a rathole`
@MitchBradley, sorry, but I would not care much if people doesn't know to interpret what they are reading. If I am looking for some information I have to read it and understand it. I would not be worried about this at all.

`Unfortunately, that level of context is rarely provided. When someone creates a new page of instructions, their typical mindset is that they are deeply immersed in what they are doing at the moment. What they write reflects that narrow focus, and then when a new user lands there, lacking any of that context, they are utterly confused.
Good documentation is hard. Really hard.`
I would suggest that before somebody is able to post to the wiki or more important to the official documentation , that some of the persons review that and give their approval. 
But there is 2 kind of documentation afaik usually in OSS projects,one is the official documentation provided by the maintainer or software developer, the other one is usually the one provided by different use cases. There is plenty of OSS out there which have this, 
and again, this imho is not really a problem. If you do not know what you are reading is all pointless.

**Comment by MitchBradley** [comment]
I'm not saying that I like FB.  In fact in many ways I really really hate it.  When Cheton created the group I cringed.  But then, eventually, I saw the very large number of CNCjs users who really do not know about anything else.  Should we just ignore them?  (Actually, I contemplate doing just that about twice a week.)

**Reply by emcniece** [reply]
The quality and audience of conversations in Facebook and Github appear a bit different. When creating an issue here in Github we request and expect deep technical detail, as it's often necessary to resolve the issue. The posts in Facebook, while some are technical, appear to be more casual and tangential.

While I am offering to help here, I can't make that same commitment to Facebook. I personally don't see any issue with letting the Facebook group be a community-driven self-help group. If anyone is feeling burnt out with offering support in any place, I'd say they should care for their own mental health first and foremost and adjust their level of commitment! As you've aptly noted Mitch, this is purely on a volunteer basis, and even if the role entailed compensation it still might not be worth the stress. 

What is the goal behind supporting the Facebook group? Is it to drive engagement, or increase adoption? And to what end - will this be a monetized product? If it's support out of a sense of obligation and not a contractual agreement, perhaps it's time to let the group grow organically 😄

**Reply by saxa** [reply]
@MitchBradley I don't care if people like or not FB, I just separate the 2 things in the relevant and less relevant, at the end we won't be able to follow everybody at every different place where the information is. I would suggest that you ignore FB requests, but don't getme wrongly, if you do care just continue doing it. No problem for me, I keep following here on github the tickets when they appear, and if the case is that I can answer or help i do it. But the thing is that usiecases for such cnc software as is cncjs is quite wide, some things are similar to others, but some are just widely distant. 

As @emcniece said, the 2 platforms and 2 kind of users who use FB or GH are very different in technical knowledge.

**Reply by cheton** [reply]
Basically, there are differences between GitHub and FB. The primary reason I created the FB group is to encourage community sharing and self-service collaboration to reduce the costs of asking questions using GitHub issues, thus we can leave real issues and discuss the roadmap on GitHub.

I have to say sorry to @MitchBradley that it causes inconveniences to you for managing the FB group for such a long time. May I suggest we post an announcement to call for volunteers on FB to manage daily posts? Ideally, we can offload the operational costs, just stay focused on important topics and avoid distractions.

**Comment by saxa** [comment]
FB is good for visibility, thats sure, nothing will probably have so much visibility elswhere. But this comes with a cost, lots of users with lots of same problems, and plus people think that some developer or maintainer has the need to solve their problem. This is how I see it.
@cheton your proposal surely will solve the issue which @MitchBradley has, and I do agree also on this, there is no need only one
person have to answer or try to solve some kind of user issues. 
I think that surely documentation need to be organized in a clear way and most probably in one place. I remember that when I started to try to use CNCjs my main problem was to find references where and how to start building and installing. It was not very difficult though, but I had to struggle a bit around.

**Comment by MitchBradley** [comment]
I was hoping that the community would support itself via FB, but I quickly discovered that the quality of community answers was just terrible.  Responders often misunderstand the question (answer a different question than was asked), give answers that are too brief to be understood, or just give wrong answers.  A lot of the questions have to do with installation problems, which require knowledge of system software, NPM, Linux techniques for autostarting programs, and other topics that very few users know anything about.

About half of the questions are really about configuring GRBL.  Many new users do not really understand that most of the difficult work is done by a separate controller running GRBL or something.  CNCjs is what they see, so they think that CNCjs is responsible for all of the behavior.

The DIY CNC world is extremely difficult to support because there are so many different components from so many different sources.  The components are inexpensive or free, with little or no support available from the suppliers.  New user finds themselves in a complex world involving mechanical systems, electronics, firmware, software, wiring, GCode, CAD/CAM, and post processors, and typically understands none of it.

One thing I learned from years of developing boot firmware for general purpose computers is "the bug is always filed against the software that displays the error message".
